### PR TITLE
Add DCGAN example

### PR DESCRIPTION
**Pull request by ptrblck**
This PR adds a DCGAN example modified from the official DCGAN PyTorch example.
`opt_levels` `O1` and `O2` were tested and create visually similar outputs.

### CODE REVIEW

File: examples/dcgan/README.md [hunk 1]
```diff
@@ -1 +1,40 @@
...
+    errD_real_scaled.backward()
+...
+with amp.scale_loss(errD_fake, optimizerD, loss_id=1) as errD_fake_scaled:
+    errD_fake_scaled.backward()
+...
+with amp.scale_loss(errG, optimizerG, loss_id=2) as errG_scaled:
+    errG_scaled.backward()
+```
+
+Note that we use different `loss_scalers` for each computed loss.
```
Comment (mcarilli) [inline examples/dcgan/README.md]: Say "This is optional, not required; see [the documentation](https://nvidia.github.io/apex/advanced.html#optionally-have-amp-use-a-different-loss-scaler-per-loss) for more detail."

File: examples/dcgan/README.md [hunk 2]
```diff
@@ -1 +1,40 @@
-Under construction...
+# Mixed Precision DCGAN Training in PyTorch
+
+`main_amp.py` is based on [https://github.com/pytorch/examples/tree/master/dcgan](https://github.com/pytorch/examples/tree/master/dcgan).
+It implements Automatic Mixed Precision (Amp) trianing of the DCGAN example for different datasets. Command-line flags forwarded to `amp.initialize` are used to easily manipulate and switch between various pure and mixed precision "optimization levels" or `opt_level`s.  For a detailed explanation of `opt_level`s, see the [updated API guide](https://nvidia.github.io/apex/amp.html).
+
+We introduce these changes to the PyTorch DCGAN example:
```
Comment (mcarilli) [inline examples/dcgan/README.md]: add ", as described in the [Multiple models/optimizers/losses](https://nvidia.github.io/apex/advanced.html#multiple-models-optimizers-losses) section of the documentation:"

File: examples/dcgan/README.md [hunk 3]
```diff
@@ -1 +1,40 @@
-Under construction...
+# Mixed Precision DCGAN Training in PyTorch
+
+`main_amp.py` is based on [https://github.com/pytorch/examples/tree/master/dcgan](https://github.com/pytorch/examples/tree/master/dcgan).
+It implements Automatic Mixed Precision (Amp) trianing of the DCGAN example for different datasets. Command-line flags forwarded to `amp.initialize` are used to easily manipulate and switch between various pure and mixed precision "optimization levels" or `opt_level`s.  For a detailed explanation of `opt_level`s, see the [updated API guide](https://nvidia.github.io/apex/amp.html).
```
Comment (mcarilli) [inline examples/dcgan/README.md]: *training

File: examples/dcgan/main_amp.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+            nn.Conv2d(ndf * 4, ndf * 8, 4, 2, 1, bias=False),
+            nn.BatchNorm2d(ndf * 8),
+            nn.LeakyReLU(0.2, inplace=True),
+            # state size. (ndf*8) x 4 x 4
+            nn.Conv2d(ndf * 8, 1, 4, 1, 0, bias=False),
+        )
+
+    def forward(self, input):
+        if input.is_cuda and self.ngpu > 1:
+            output = nn.parallel.data_parallel(self.main, input, range(self.ngpu))
```
Comment (mcarilli) [inline examples/dcgan/main_amp.py]: Does ngpus > 1 with this syntax work with Amp?  DataParallel is usually not recommended, although what's used here is a functional version rather than a module wrapper.

File: examples/dcgan/main_amp.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+            nn.ReLU(True),
+            # state size. (ngf) x 32 x 32
+            nn.ConvTranspose2d(    ngf,      nc, 4, 2, 1, bias=False),
+            nn.Tanh()
+            # state size. (nc) x 64 x 64
+        )
+
+    def forward(self, input):
+        if input.is_cuda and self.ngpu > 1:
+            output = nn.parallel.data_parallel(self.main, input, range(self.ngpu))
```
Comment (mcarilli) [inline examples/dcgan/main_amp.py]: Does ngpus > 1 with this syntax work with Amp? DataParallel is usually not recommended, although what's used here is a functional version rather than a module wrapper.  If it works, I'm happy with it.

Comment (ptrblck) [reply examples/dcgan/main_amp.py]: I just ran a test using 8 GPUs and the outputs are visually similar, although we might see some more mode collapses.

Comment (mcarilli) [reply examples/dcgan/main_amp.py]: Sounds good to me.  We can add DDP support if we decide it's useful later.
